### PR TITLE
ENH: Multi-level join on multi-indexes

### LINKED ISSUE #6360: ENH: merge multi-index with a multi-index
#3662 is about merging a single-level index with a mi

This is is about a multi-multi merge

```
# example described there:
In [11]: df = pd.DataFrame(np.random.randn(10,5))
    ...: df1 = df[range(0, 3)].set_index([0, 2])
    ...: df2 = df[range(2, 5)].set_index([2, 4])
    ...: res = df1.join(df2, how='inner')  # empty. Do I need to specify something here?
    ...: exp = pd.DataFrame({1: df[1], 3: df[3]})
    ...: exp.index = df[2]
```

### PR DESCRIPTION
**Pull request by harisbal**
closes #6360

Allow for merging on multiple levels of multi-indexes

### CODE REVIEW

File: pandas/tests/reshape/test_merge.py [hunk 1]
```diff
@@ -1136,14 +1136,14 @@ def test_join_multi_levels(self):
 
         def f():
             household.join(portfolio, how='inner')
-        pytest.raises(ValueError, f)
+        self.assertRaises(ValueError, f)
```
Comment (gfyoung) [inline pandas/tests/reshape/test_merge.py]: Do not change this.  We're using `pytest` and not `nosetest`.

File: pandas/tests/reshape/test_merge.py [hunk 2]
```diff
@@ -1136,14 +1136,14 @@ def test_join_multi_levels(self):
...
 
         portfolio2 = portfolio.copy()
         portfolio2.index.set_names(['household_id', 'foo'])
 
         def f():
             portfolio2.join(portfolio, how='inner')
-        pytest.raises(ValueError, f)
+        self.assertRaises(ValueError, f)
```
Comment (gfyoung) [inline pandas/tests/reshape/test_merge.py]: Same as <a href="https://github.com/pandas-dev/pandas/pull/16162/files#r114061443">here</a>.

File: pandas/tests/reshape/test_merge.py [hunk 3]
```diff
@@ -1208,12 +1204,179 @@ def f():
...
+                     Period=['AM','AM', 'PM'],
+                     LinkType=['a', 'b', 'a'],
+                     Distance=[100, 110, 120]),
+                columns=['Origin', 'Destination', 'Period', 
+                         'LinkType', 'Distance'])
+            .set_index(['Origin', 'Destination','Period', 'LinkType']))
+                
+        def f():
+            matrix.join(distances2, how='left')
+        self.assertRaises(TypeError, f)
```
Comment (gfyoung) [inline pandas/tests/reshape/test_merge.py]: Same as <a href="https://github.com/pandas-dev/pandas/pull/16162/files#r114061443">here</a>.

File: pandas/tests/reshape/test_merge.py [hunk 4]
```diff
@@ -1208,12 +1204,179 @@ def f():
...
+                     Per=['AM','PM','IP','AM','OP','IP', 'AM'],
+                     LinkTyp=['a', 'a', 'c', 'b', 'a', 'b', 'a'],
+                     Dist=[100, 80, 90, 80, 75, 35, 55]),
+                columns=['Orig', 'Dest', 'Per', 
+                         'LinkTyp', 'Dist'])
+            .set_index(['Orig', 'Dest','Per', 'LinkTyp']))
+        
+        def f():
+            matrix.join(distances2, how='left')
+        self.assertRaises(ValueError, f)
```
Comment (gfyoung) [inline pandas/tests/reshape/test_merge.py]: Same as <a href="https://github.com/pandas-dev/pandas/pull/16162/files#r114061443">here</a>.

**Comment by harisbal** [comment]
Can anyone give some insight on the failure of Travis CI tests?

**Comment by TomAugspurger** [comment]
Some linting (style) issues, starting at https://travis-ci.org/pandas-dev/pandas/jobs/232494137#L1403

You can `pip install flake8` and run it on the files your changing.

File: pandas/core/indexes/base.py [hunk 5]
```diff
@@ -3028,27 +3028,84 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
+                new_levels = join_index.levels
+                new_labels = join_index.labels
+                new_names = join_index.names
+
+            join_index = MultiIndex(levels=new_levels, labels=new_labels,
+                                    names=new_names, verify_integrity=False)
+
+            return join_index, lidx, ridx
+
+        else:
```
Comment (jreback) [inline pandas/core/indexes/base.py]: you don't need the else here

File: pandas/tests/reshape/test_merge.py [hunk 6]
```diff
@@ -1195,7 +1191,7 @@ def f():
         expected = (
             DataFrame(dict(
                 household_id=[1, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4],
-                asset_id=["nl0000301109", "nl0000289783", "gb00b03mlx29",
+                asset_id=["nl0000301109", "nl0000301109", "gb00b03mlx29",
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: why is this changed?

File: pandas/tests/reshape/test_merge.py [hunk 7]
```diff
@@ -1208,11 +1204,179 @@ def f():
...
+            .set_index(["household_id", "asset_id", "t"])
+            .reindex(columns=['share', 'log_return']))
+
+        result = (merge(household.reset_index(), log_return.reset_index(),
+                  on=['asset_id'], how='outer')
+                  .set_index(['household_id', 'asset_id', 't']))
+
+        assert_frame_equal(result, expected)
+
+    def test_join_multi_levels3(self):
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: where are the error cases, e.g. mistmatch on multi-levels? misnamed levels.

File: pandas/core/indexes/base.py [hunk 8]
```diff
@@ -3028,27 +3028,84 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
-                                      "implemented")
-        jl = overlap[0]
+            raise ValueError("cannot join with no overlapping index names")
+
+        if self_is_mi and other_is_mi:
+            self_tmp = self.droplevel(ldrop_levels)
+            other_tmp = other.droplevel(rdrop_levels)
+
+            if not (other_tmp.is_unique and self_tmp.is_unique):
+                raise TypeError(" The index resulting from the overlapping "
```
Comment (jreback) [inline pandas/core/indexes/base.py]: need testing for this

File: pandas/core/indexes/base.py [hunk 9]
```diff
@@ -3028,27 +3028,84 @@ def join(self, other, how='left', level=None, return_indexers=False,
 
     def _join_multi(self, other, how, return_indexers=True):
         from .multi import MultiIndex
-        self_is_mi = isinstance(self, MultiIndex)
-        other_is_mi = isinstance(other, MultiIndex)
+
+        def _complete_join():
```
Comment (jreback) [inline pandas/core/indexes/base.py]: this should be a separate method, too confusing here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reshape/test_merge.py (hunk 6 above)
Comment (harisbal) [inline pandas/tests/reshape/test_merge.py]: I believe it was a mistake in the first instance. The household dataframe in this method suggests this correction

File: pandas/core/indexes/base.py (hunk 8 above)
Comment (harisbal) [inline pandas/core/indexes/base.py]: https://github.com/harisbal/pandas/blob/4ed307074a4c376db209b44aae5b7eef00276d26/pandas/tests/reshape/test_merge.py#L1323-L1337

File: pandas/tests/reshape/test_merge.py (hunk 7 above)
Comment (harisbal) [inline pandas/tests/reshape/test_merge.py]: https://github.com/harisbal/pandas/blob/4ed307074a4c376db209b44aae5b7eef00276d26/pandas/tests/reshape/test_merge.py#L1339-L1352

File: pandas/tests/reshape/test_merge.py (hunk 6 above)
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: why does the existing code not fail then? before this change

Comment (harisbal) [inline pandas/tests/reshape/test_merge.py]: It was raising a NotImplementedError

**Review by harisbal — COMMENTED** [review]
Changes implemented

File: pandas/core/indexes/base.py [hunk 10]
```diff
@@ -3028,27 +3028,61 @@ def join(self, other, how='left', level=None, return_indexers=False,
 
     def _join_multi(self, other, how, return_indexers=True):
         from .multi import MultiIndex
-        self_is_mi = isinstance(self, MultiIndex)
-        other_is_mi = isinstance(other, MultiIndex)
 
         # figure out join names
         self_names = [n for n in self.names if n is not None]
         other_names = [n for n in other.names if n is not None]
         overlap = list(set(self_names) & set(other_names))
```
Comment (jreback) [inline pandas/core/indexes/base.py]: this should remain a set

File: pandas/core/indexes/base.py [hunk 11]
```diff
@@ -3028,27 +3028,61 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
+        other_is_mi = isinstance(other, MultiIndex)
+
+        if self_is_mi and other_is_mi:
+            self_tmp = self.droplevel(ldrop_lvls)
+            other_tmp = other.droplevel(rdrop_lvls)
 
-        # make the indices into mi's that match
-        if not (self_is_mi and other_is_mi):
+            if not (self_tmp.is_unique and other_tmp.is_unique):
+                raise TypeError(" The index resulting from the overlapping "
```
Comment (jreback) [inline pandas/core/indexes/base.py]: use the original error message and it should be a ValueError

File: pandas/core/indexes/base.py [hunk 12]
```diff
@@ -3028,27 +3028,61 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
+            dropped_lvls = ldrop_lvls + rdrop_lvls
+
+            join_index, lidx, ridx = self_tmp.join(other_tmp, how=how,
+                                                   return_indexers=True)
+
+            if how == 'left':
+                join_index = self
+            elif how == 'right':
+                join_index = other
+            else:
```
Comment (jreback) [inline pandas/core/indexes/base.py]: else is unecessary

File: pandas/core/indexes/base.py [hunk 13]
```diff
@@ -3028,27 +3028,61 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
+                new_lvls = join_index.levels
+                new_lbls = join_index.labels
+                new_nms = join_index.names
+
+            join_index = MultiIndex(levels=new_lvls, labels=new_lbls,
+                                    names=new_nms, verify_integrity=False)
+
+            return join_index, lidx, ridx
+        else:
+            jl = overlap[0]
```
Comment (jreback) [inline pandas/core/indexes/base.py]: don't need the else here as you are returning

File: pandas/core/indexes/base.py [hunk 14]
```diff
@@ -3028,27 +3028,61 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
-                                      "overlap on a multi-index is not "
-                                      "implemented")
-        jl = overlap[0]
+            raise ValueError("cannot join with no overlapping index names")
+
+        # Drop the non matching levels
+        ldrop_lvls = [l for l in self_names if l not in overlap]
+        rdrop_lvls = [l for l in other_names if l not in overlap]
+
+        self_is_mi = isinstance(self, MultiIndex)
```
Comment (jreback) [inline pandas/core/indexes/base.py]: instead of doing this inline it should be a separate method, its too long

File: pandas/core/indexes/base.py [hunk 15]
```diff
@@ -3065,9 +3099,30 @@ def _join_multi(self, other, how, return_indexers=True):
                     return result[0], result[2], result[1]
             return result
 
-        # 2 multi-indexes
-        raise NotImplementedError("merging with both multi-indexes is not "
-                                  "implemented")
+    def _complete_multi_join(self, other, join_idx, lidx, ridx, dropped_lvls):
+        new_lvls = join_idx.levels
```
Comment (jreback) [inline pandas/core/indexes/base.py]: doc-string on what this is

File: pandas/core/indexes/base.py [hunk 16]
```diff
@@ -3065,9 +3099,30 @@ def _join_multi(self, other, how, return_indexers=True):
...
+        new_lbls = join_idx.labels
+        new_nms = join_idx.names
+
+        for n in dropped_lvls:
+            if n in self.names:
+                idx = lidx
+                lvls = self.levels[self.names.index(n)].values
```
Comment (jreback) [inline pandas/core/indexes/base.py]: don't abbreviate on the variable names, use new_levels and new_labels etc

File: pandas/tests/reshape/test_merge.py [hunk 17]
```diff
@@ -1208,11 +1204,179 @@ def f():
...
+                     Distance=[100, 80, 90, 80, np.nan, 75, 35, 55]),
+                columns=['Origin', 'Destination', 'Period', 'TripPurp',
+                         'LinkType', 'Trips', 'Distance'])
+            .set_index(['Origin', 'Destination', 'Period',
+                        'TripPurp', 'LinkType']))
+
+        result = matrix.join(distances, how='outer')
+        assert_frame_equal(result, expected)
+
+        # Non-unique resulting index
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: separate test for non-uniques

File: pandas/tests/reshape/test_merge.py [hunk 18]
```diff
@@ -1208,11 +1204,179 @@ def f():
...
+                         'LinkType', 'Distance'])
+            .set_index(['Origin', 'Destination', 'Period', 'LinkType']))
+
+        expected = (
+            pd.DataFrame(
+                dict(Origin=[1, 1, 2, 2, 3],
+                     Destination=[1, 2, 1, 3, 1],
+                     Period=['AM', 'PM', 'IP', 'AM', 'OP'],
+                     TripPurp=['hbw', 'nhb', 'hbo', 'nhb', 'hbw'],
+                     Trips=[1987, 3647, 2470, 4296, 4444],
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: in fact why don't you move all of these added tests to a new TestMultiMulti class for testing (and other relevant tests if any).

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
pls rebase and update

**Review by harisbal — COMMENTED** [review]
Rebased

**Review by TomAugspurger — APPROVED** [review]
cc @jreback

File: pandas/core/indexes/base.py [hunk 19]
```diff
@@ -3111,46 +3111,125 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
+
+        self_is_mi = isinstance(self, MultiIndex)
+        other_is_mi = isinstance(other, MultiIndex)
+
+        if self_is_mi and other_is_mi:
+            self_tmp = self.droplevel(ldrop_levels)
+            other_tmp = other.droplevel(rdrop_levels)
+
+            if not (self_tmp.is_unique and other_tmp.is_unique):
+                raise ValueError("Join on level between two MultiIndex objects"
```
Comment (jreback) [inline pandas/core/indexes/base.py]: test for this?

File: pandas/core/indexes/base.py [hunk 20]
```diff
@@ -3111,46 +3111,125 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
+                                                                  lidx, ridx,
+                                                                  dropped_lvls)
+
+            else:
+                if how == 'left':
+                    join_index = self
+                elif how == 'right':
+                    join_index = other
+
+                levels = join_index.levels
```
Comment (jreback) [inline pandas/core/indexes/base.py]: move these 3 assignments after the .join, then you can make the 'how' an if/elif (for each of outer/left/right)

File: pandas/core/indexes/base.py [hunk 21]
```diff
@@ -3111,46 +3111,125 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
-
-            level = other.names.index(jl)
-            result = self._join_level(other, level, how=how,
-                                      return_indexers=return_indexers)
-
-            if flip_order:
-                if isinstance(result, tuple):
-                    return result[0], result[2], result[1]
-            return result
+        flip_order = False
```
Comment (jreback) [inline pandas/core/indexes/base.py]: huh?

if you need to do something based on the how, then do it above

File: pandas/core/indexes/base.py [hunk 22]
```diff
@@ -3111,46 +3111,125 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
+            self, other = other, self
+            flip_order = True
+            # flip if join method is right or left
+            how = {'right': 'left', 'left': 'right'}.get(how, how)
+
+        level = other.names.index(jl)
+        result = self._join_level(other, level, how=how,
+                                  return_indexers=return_indexers)
+
+        if flip_order:
```
Comment (jreback) [inline pandas/core/indexes/base.py]: this is non-intuitive here, have no idea what is going one

File: pandas/core/indexes/base.py [hunk 23]
```diff
@@ -3111,46 +3111,125 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
+            if isinstance(result, tuple):
+                return result[0], result[2], result[1]
+        return result
+
+    def _complete_outer_join(self, other, join_idx, lidx, ridx,
```
Comment (jreback) [inline pandas/core/indexes/base.py]: move to this pandas/core/reshape/merge.py (below _get_join_indexers) and just import where you need

File: pandas/tests/reshape/test_merge.py [hunk 24]
```diff
@@ -1344,11 +1341,191 @@ def f():
...
+                  on=['asset_id'], how='outer')
+                  .set_index(['household_id', 'asset_id', 't']))
+
+        assert_frame_equal(result, expected)
+
+    def test_join_multi_levels3(self):
+        # Multi-index join tests
+        # Self join
+        matrix = (
+            pd.DataFrame(
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: pls parametrize or use fixtures for this.

File: pandas/tests/reshape/test_merge.py [hunk 25]
```diff
@@ -1344,11 +1341,191 @@ def f():
...
+            .set_index(["household_id", "asset_id", "t"])
+            .reindex(columns=['share', 'log_return']))
+
+        result = (merge(household.reset_index(), log_return.reset_index(),
+                  on=['asset_id'], how='outer')
+                  .set_index(['household_id', 'asset_id', 't']))
+
+        assert_frame_equal(result, expected)
+
+    def test_join_multi_levels3(self):
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: rename to something more useful

**Review by jreback — CHANGES_REQUESTED** [review]
a number of comments

**Comment by jreback** [comment]
also the docs in merging.rst need updating. pls add a small sub-section in 0.21.0 as well showcasing the new functionaility.

**Comment by harisbal** [comment]
Thank you @jreback, I'll implement the changes

**Comment by jreback** [comment]
@harisbal great. sorry it took so long on this.

**Comment by harisbal** [comment]
It was actually my fault. Thanks again for all the comments :)

File: pandas/core/indexes/base.py (hunk 21 above)
Comment (harisbal) [reply pandas/core/indexes/base.py]: The part from `jl = list(overlap)[0]` and onwards refers to cases where not both indexes are multiindexes.
This is legacy code that I didn't modify. Shall I try to rewrite?

File: pandas/core/indexes/base.py (hunk 22 above)
Comment (harisbal) [reply pandas/core/indexes/base.py]: Same as above. Legacy code that I didn't modify

File: pandas/core/reshape/merge.py [hunk 26]
```diff
@@ -1066,6 +1066,54 @@ def _get_join_indexers(left_keys, right_keys, sort=False, how='inner',
     return join_func(lkey, rkey, count, **kwargs)
 
 
+def _complete_outer_join(self, other, join_idx, lidx, ridx, dropped_levels):
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: call these left/right (not self, other), add to doc-string

File: pandas/core/reshape/merge.py [hunk 27]
```diff
@@ -1066,6 +1066,54 @@ def _get_join_indexers(left_keys, right_keys, sort=False, how='inner',
...
+    """
+    *this is an internal non-public method*
+
+    Complete the index in case of outer join
+
+    Parameters
+    ----------
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: document like
```
join_index : Index
    the index of the join between common leves
lidx : intp array
    left indexer
....
```

etc

File: pandas/core/indexes/base.py [hunk 28]
```diff
@@ -3111,27 +3111,64 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
+
+            if not (self_tmp.is_unique and other_tmp.is_unique):
+                raise ValueError("Join on level between two MultiIndex objects"
+                                 "is ambiguous")
+
+            # Append to the returned Index the non-overlapping levels
+            dropped_lvls = ldrop_levels + rdrop_levels
+
+            # tmp_index is equivalent of index when how='inner'
+            tmp_index, lidx, ridx = self_tmp.join(other_tmp, how=how,
```
Comment (jreback) [inline pandas/core/indexes/base.py]: don't use ``tmp_index`` call this something else (and self_tmp)

File: pandas/core/indexes/base.py [hunk 29]
```diff
@@ -3111,27 +3111,64 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
+            other_tmp = other.droplevel(rdrop_levels)
+
+            if not (self_tmp.is_unique and other_tmp.is_unique):
+                raise ValueError("Join on level between two MultiIndex objects"
+                                 "is ambiguous")
+
+            # Append to the returned Index the non-overlapping levels
+            dropped_lvls = ldrop_levels + rdrop_levels
+
+            # tmp_index is equivalent of index when how='inner'
```
Comment (jreback) [inline pandas/core/indexes/base.py]: explicity enumerate all of the how cases

```
if how == 'innter':
     pass
elif how == 'outer':
    # outer_join
elif how == 'left'
.....
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py (hunk 28 above)
Comment (harisbal) [reply pandas/core/indexes/base.py]: How about:
self_tmp --> left
other_tmp --> right
tmp_index --> inner_join_index

File: doc/source/merging.rst [hunk 30]
```diff
@@ -1120,6 +1120,13 @@ This is not Implemented via ``join`` at-the-moment, however it can be done using
           labels=['left', 'right'], vertical=False);
    plt.close('all');
 
+For previous versions can be done using the following.
```
Comment (jreback) [inline doc/source/merging.rst]: rather I would say this is equivalent to

File: doc/source/merging.rst [hunk 31]
```diff
@@ -1098,7 +1098,8 @@ This is equivalent but less verbose and more memory efficient / faster than this
 Joining with two multi-indexes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-This is not Implemented via ``join`` at-the-moment, however it can be done using the following.
+.. versionadded:: 0.21.0
+As of version 0.21.0 joining on two multi-indexes is possible:
```
Comment (jreback) [inline doc/source/merging.rst]: can you add a similar example to a sub-section in whatsnew (and put a ref to here). you can use this example and show how we *could* do it previously (e.g. the below section), and how it will just work now.

File: pandas/core/indexes/base.py [hunk 32]
```diff
@@ -3111,27 +3111,53 @@ def join(self, other, how='left', level=None, return_indexers=False,
...
+                                                              join_index,
+                                                              lidx, ridx)
+
+            multilevel_join_index = MultiIndex(levels=levels, labels=labels,
+                                               names=names,
+                                               verify_integrity=False)
+
+            return multilevel_join_index, lidx, ridx
+
         if not (self_is_mi and other_is_mi):
```
Comment (jreback) [inline pandas/core/indexes/base.py]: I think this just becomes an else here

Comment (jreback) [reply pandas/core/indexes/base.py]: or you can just remove the if as well (IOW we have covered both cases).

File: pandas/core/reshape/merge.py [hunk 33]
```diff
@@ -1066,6 +1066,81 @@ def _get_join_indexers(left_keys, right_keys, sort=False, how='inner',
...
+    join_idx : Index
+        the index of the join between the common levels of left and right
+    how : {'left', 'right', 'outer', 'inner'}
+    lidx : intp array
+        left indexer
+    right : intp array
+        right indexer
+    dropped_levels : str array
+        list of non-common levels
+    Returns
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: blank line before Returns

File: pandas/core/reshape/merge.py [hunk 34]
```diff
@@ -1066,6 +1066,81 @@ def _get_join_indexers(left_keys, right_keys, sort=False, how='inner',
...
+        levels of combined multiindexes
+    labels : str array
+        labels of combined multiindexes
+    names : str array
+        names of combined multiindexes
+
+    """
+
+    if how == 'outer':
+        levels = join_idx.levels
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: maybe a few comments on what you are doing here (for the outer case), IOW why iterating over the dropped levels, etc.

File: pandas/tests/reshape/test_merge.py [hunk 35]
```diff
@@ -1344,32 +1341,194 @@ def f():
...
+                                               ('inner', expected_innerj),
+                                               ('outer', expected_outerj)])
+    def test_join_multi_multi(self, how, expected):
+        # Multi-index join tests
+        left = self.left()
+        right = self.right()
+
+        result = left.join(right, how=how)
+        tm.assert_frame_equal(result, expected)
+
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: is it possible to generically also tests the equivalency (this is just a copy-paste from above, obviously have to adapt here)

```
result = (merge(household.reset_index(), log_return.reset_index(),
                          on=['asset_id'], how='inner')
                    .set_index(['household_id', 'asset_id', 't']))
```

File: pandas/tests/reshape/test_merge.py [hunk 36]
```diff
@@ -1344,32 +1341,194 @@ def f():
...
+        result = (merge(household.reset_index(), log_return.reset_index(),
+                  on=['asset_id'], how='outer')
+                  .set_index(['household_id', 'asset_id', 't']))
+
+        assert_frame_equal(result, expected)
+
+
+class TestJoinMultiMulti(object):
+    @pytest.fixture
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: I would simply put the fixture outside the class, right above TestJoinMultiMulti, then you can just pass them in as needed

File: pandas/tests/reshape/test_merge.py [hunk 37]
```diff
@@ -1344,32 +1341,194 @@ def f():
...
+            columns=['Origin', 'Destination', 'Period', 'TripPurp',
+                     'LinkType', 'Trips', 'Distance'])
+        .set_index(['Origin', 'Destination', 'Period',
+                    'TripPurp', 'LinkType']))
+
+    @pytest.mark.parametrize('how, expected', [('left', expected_leftj),
+                                               ('right', expected_rightj),
+                                               ('inner', expected_innerj),
+                                               ('outer', expected_outerj)])
+    def test_join_multi_multi(self, how, expected):
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: e.g. you pass left, right in here

File: pandas/tests/reshape/test_merge.py [hunk 38]
```diff
@@ -1344,32 +1341,194 @@ def f():
...
+                 LinkType=['a', 'a', 'c', 'b', np.nan, 'a', 'b', 'a'],
+                 Trips=[1987, 3647, 2470, np.nan,
+                        4296, 4444, np.nan, np.nan],
+                 Distance=[100, 80, 90, 80, np.nan, 75, 35, 55]),
+            columns=['Origin', 'Destination', 'Period', 'TripPurp',
+                     'LinkType', 'Trips', 'Distance'])
+        .set_index(['Origin', 'Destination', 'Period',
+                    'TripPurp', 'LinkType']))
+
+    @pytest.mark.parametrize('how, expected', [('left', expected_leftj),
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: these you actually DO need to evaluate (e.g. ``expected_left()``) because these are simply calling a function and are not 'parameters' (if they are in the signature they are parameters).

File: pandas/tests/reshape/test_merge.py [hunk 39]
```diff
@@ -1408,28 +1567,27 @@ def test_join_multi_dtypes(self, d1, d2):
...
-    return DataFrame(
-        {'X': Series(['foo', 'bar']).astype('category',
-                                            categories=['foo', 'bar']),
-         'Z': [1, 2]})
-
-
 class TestMergeCategorical(object):
 
-    def test_identical(self, left):
+    @pytest.fixture
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: would leave these where they were

File: pandas/tests/reshape/test_merge.py [hunk 40]
```diff
@@ -1439,9 +1597,11 @@ def test_identical(self, left):
                           index=['X', 'Y_x', 'Y_y'])
         assert_series_equal(result, expected)
 
-    def test_basic(self, left, right):
```
Comment (jreback) [inline pandas/tests/reshape/test_merge.py]: so just leave this, it is very typical, writing ``left = self.left()`` is not.

**Review by jreback — CHANGES_REQUESTED** [review]
can you add some tests versus empty frames (matching levels but empty). test all the hows's

**Comment by harisbal** [comment]
@jreback I have implemented the changes but one test is failing and I'm struggling to understand why. Shall I upload so you can give it a look?

File: doc/source/merging.rst (hunk 31 above)
Comment (harisbal) [reply doc/source/merging.rst]: Shall I wait to include it in 0.21.1?

Comment (jreback) [reply doc/source/merging.rst]: going to be for next version (so you can wait to move it is ok), after we tag can merge shortly after.

**Comment by jreback** [comment]
@harisbal what test is failing? l

**Comment by jreback** [comment]
can you rebase

**Comment by harisbal** [comment]
@jreback  Any idea why the tests do not fail when I run them locally?

**Comment by TomAugspurger** [comment]
Hmm @harisbal seems like there was a rebase snafu in https://github.com/pandas-dev/pandas/pull/16162/commits/5c27d875759f265bf79d7c9c989cf2464cab8a59. Need help fixing it up?

**Comment by harisbal** [comment]
@TomAugspurger  thank you very much but I managed to rebase properly. I now have an issue when I try to outer join on an empty level. I cannot debug after a certain point (I assume cython comes into play) so some help would be much appreciated

**Comment by jreback** [comment]
can you rebase. let's see if we can get this in.

**Comment by harisbal** [comment]
Two tests fail and I can't see how to fix the issue

**Comment by auvipy** [comment]
rebase and sqash commits or create a new pr

**Comment by jreback** [comment]
@harisbal sorry let this drop. can you rebase and we can get this in.

**Comment by harisbal** [comment]
This pull request is dirty beyond repair. I'll squash and start a new one
